src/sinks/apex/mod.rs

```diff
@@ -0,0 +1,170 @@
+use bytes::Bytes;
+use futures::FutureExt;
+use futures_util::SinkExt;
+use http::{Request, StatusCode, Uri};
+use hyper::Body;
+use serde::{Deserialize, Serialize};
+use serde_json::json;
+
+#[cfg(feature = "apex-integration-tests")]
+#[cfg(test)]
+mod integration_tests;
+
+use crate::{
+    config::{
+        AcknowledgementsConfig, GenerateConfig, Input, SinkConfig, SinkContext, SinkDescription,
+    },
+    event::Event,
+    http::HttpClient,
+    sinks::util::{
+        encoding::Transformer,
+        http::{BatchedHttpSink, HttpEventEncoder, HttpSink},
+        BatchConfig, BoxedRawValue, JsonArrayBuffer, SinkBatchSettings, TowerRequestConfig,
+        UriSerde,
+    },
+};
+
+#[derive(Deserialize, Serialize, Clone, Debug)]
+#[serde(deny_unknown_fields)]
+pub struct ApexSinkConfig {
+    uri: UriSerde,
+    project_id: String,
+    api_token: String,
+
+    #[serde(default)]
+    batch: BatchConfig<ApexDefaultBatchSettings>,
+
+    #[serde(default)]
+    request: TowerRequestConfig,
+
+    #[serde(
+        default,
+        deserialize_with = "crate::serde::bool_or_struct",
+        skip_serializing_if = "crate::serde::skip_serializing_if_default"
+    )]
+    acknowledgements: AcknowledgementsConfig,
+}
+
+#[derive(Clone, Copy, Debug, Default)]
+struct ApexDefaultBatchSettings;
+
+impl SinkBatchSettings for ApexDefaultBatchSettings {
+    const MAX_EVENTS: Option<usize> = None;
+    const MAX_BYTES: Option<usize> = None;
+    const TIMEOUT_SECS: f64 = 1.0;
+}
```

Comment:
To be honest, this strikes me as a bit too permissive.

Does Apex really not any limits on the payload size for ingesting events? This should likely be removed and `RealtimeSizeBasedDefaultBatchSettings` used instead unless there's a particularly good reason not to use... especially since the sink has no compression support as-is.

Reply:
I've not discovered any limits either in the documentation or in testing, but seeing as it's a product that you self-host, that's not terribly surprising to me.

Either way, I'm happy to use `RealtimeSizeBasedDefaultBatchSettings`!